### DISCUSSION TITLE
Catia wrl nesting

### DISCUSSION THREAD

**Discussion by YamaPrograms** [fel88/DeepNestPort#30] [Ideas]
Hello how can I get contact with you I have some questions. I working on catia wrl nesting in c# winform. I stuck I got my limit points from wrl file but I don't know what should I do next. From which file is nesting algorithm in you project? I'm new in this GitHub etc stuff :)

**Comment by fel88** [comment]
Hi! We can discuss everything you want about this project right here. Just give me a questions list and I'll try to answer

**Reply by YamaPrograms** [reply]
I'm trying to make nesting from .wrl file I save it from catiav5. But how can I create an algorithm like yours? I take points from wrl file then  creating lines with my code. Btw using c# winform with catia v5 references. Here is my code but I'm confused about next steps...  I can send my files if its ok for you? How can I send it :D

`private List<(double, double)> ReadCoordinatesFromWrl(string filePath)
{
    List<(double, double)> coordinates = new List<(double, double)>();

    try
    {
        // Dosyayı satır satır oku
        var lines = System.IO.File.ReadAllLines(filePath);

        bool isInsidePointBlock = false;

        foreach (var line in lines)
        {
            var trimmedLine = line.Trim();

            // "point" bloğunu bulmaya çalışıyoruz
            if (trimmedLine.StartsWith("point"))
            {
                isInsidePointBlock = true;
                continue; // point kısmını geç
            }

            // Eğer point bloğu içerisindeysek
            if (isInsidePointBlock)
            {
                // Eğer satır kapama işaretleri içeriyorsa, bu bloğun sonuna gelmişizdir
                if (trimmedLine.Contains("]"))
                {
                    isInsidePointBlock = false;
                    continue;
                }

                // Satırı al ve noktaları ayıkla
                string[] parts = trimmedLine.Split(new[] { ',', ' ' }, StringSplitOptions.RemoveEmptyEntries);

                for (int i = 0; i < parts.Length; i += 3) // Her üç eleman bir noktayı temsil eder
                {
                    try
                    {
                        double x = double.Parse(parts[i]);
                        double y = double.Parse(parts[i + 1]);
                        // Z'yi yok sayıyoruz (parts[i + 2])

                        coordinates.Add((x, y));
                    }
                    catch (Exception ex)
                    {
                        Debug.WriteLine($"Koordinat okunurken hata oluştu: {ex.Message}");
                    }
                }
            }
        }
    }
    catch (Exception ex)
    {
        MessageBox.Show($"Dosya okunurken hata oluştu: {ex.Message}");
    }

    return coordinates;
}`

`private void WRLBtn_Click(object sender, EventArgs e)
{
    // WRL dosyasından koordinatları oku
    string wrlFilePath = "C:\\Users\\Mert\\Desktop\\asdf\\T.wrl"; // Buraya WRL dosyasının yolunu yaz
    var coordinates = ReadCoordinatesFromWrl(wrlFilePath);

    // Eğer koordinatlar bulunamadıysa uyarı ver
    if (coordinates == null || coordinates.Count == 0)
    {
        MessageBox.Show("WRL dosyasından koordinatlar okunamadı.");
        return;
    }

    // CATIA'ya bağlan
    catiaApp = (INFITF.Application)Marshal.GetActiveObject("CATIA.Application");

    // Yeni bir ürün (Product) oluştur
    Documents catDocuments = catiaApp.Documents;
    ProductDocument productDoc = (ProductDocument)catDocuments.Add("Product");
    mainProduct = productDoc.Product;
    mainProduct.set_PartNumber("Nesting_Product");

    // Bu kısımda part oluşturma işlemi var ve oluşturulan part ana product'a eklenecek
    PartDocument partDoc = (PartDocument)catDocuments.Add("Part");
    Part part = partDoc.Part;

    Product limitPartProduct = partDoc.Product;
    limitPartProduct.set_PartNumber("Limit_Part");

    // Sketches listesini al
    Bodies bodies = part.Bodies;
    MECMOD.Body body = bodies.Item("PartBody");
    Sketches sketches = body.Sketches;

    // XY düzlemine referans oluştur
    OriginElements originElements = part.OriginElements;
    Reference planeXY = (Reference)originElements.PlaneXY;

    // Sketch oluştur
    Sketch sketch = sketches.Add(planeXY);

    // 2D Factory başlat
    Factory2D factory2D = sketch.OpenEdition();

    // WRL dosyasından alınan koordinatlara göre noktalar oluştur
    foreach (var coord in coordinates)
    {
        double xValue = coord.Item1;
        double yValue = coord.Item2;

        // Her bir noktanın koordinatını loglayalım
        Debug.WriteLine($"Nokta: X = {xValue}, Y = {yValue}");

        // Noktayı oluştur
        Point2D point = factory2D.CreatePoint(xValue, yValue);
    }

    // WRL dosyasından alınan koordinatlara göre çizgiler oluştur
    for (int i = 0; i < coordinates.Count - 1; i++)
    {
        double x1 = coordinates[i].Item1;
        double y1 = coordinates[i].Item2;
        double x2 = coordinates[i + 1].Item1;
        double y2 = coordinates[i + 1].Item2;

        // 0, 0 noktasını atla
        if ((x1 == 0 && y1 == 0) || (x2 == 0 && y2 == 0))
        {
            continue; // Eğer noktalardan biri 0,0 ise bu çizgiyi atla
        }

        // Noktaların koordinatlarını loglayalım
        Debug.WriteLine($"Çizgi: Başlangıç = ({x1}, {y1}), Bitiş = ({x2}, {y2})");

        // Noktalar arasındaki çizgiyi oluştur
        Point2D point1 = factory2D.CreatePoint(x1, y1);
        Point2D point2 = factory2D.CreatePoint(x2, y2);

        // İki nokta arasında çizgi oluştur
        Line2D line = factory2D.CreateLine(x1, y1, x2, y2);
        line.StartPoint = point1;
        line.EndPoint = point2;
    }

    // Sketche işlemi kapat ve parçayı güncelle
    sketch.CloseEdition();
    part.Update();

    // Oluşturulan part'ı ana product'a ekle
    mainProduct.Products.AddComponent(partDoc.Product);

    // Parçayı ana product'a ekledikten sonra kapat
    partDoc.Close();

    MessageBox.Show($"Toplam {coordinates.Count} nokta başarıyla eklendi.");
}`

**Reply by fel88** [reply]
Please, take a look at DeepNestConsole.Core project . I believe this is exactly what you need and it can be easily adapted to your task. It doesn't have UI interface, just a nesting loop.

**Reply by xenc0d3r** [reply]
Is there anyway to integrate deepnest into my working electron js app. Or how can we use deepnest functionality through a backend ? Thank you.

**Reply by fel88** [reply]
Hi, this project is just a C# port of the original Electron/JS project (DeepNest). So, you might be looking for something else. Anyway, you can try the console version (https://github.com/fel88/DeepNestPort/tree/master/Core/DeepNestConsole.Core) maybe it'll help you somehow.

**Reply by xenc0d3r** [reply]
Thank you for fast reply. 
Is console app capable of same functionality as deepnest backend.?

**Reply by fel88** [reply]
Yep, sure. They both (console and GUI) use the same library [DeepNestLib.Core](https://github.com/fel88/DeepNestPort/tree/master/Core/DeepNestLib.Core)

**Reply by xenc0d3r** [reply]
Thank you so much.
